### PR TITLE
Make Response.content() return None for an empty response

### LINKED ISSUE #3698: AttributeError: 'NoneType' object has no attribute 'read'
Hello :)

After a recent upgrade for our [coala](https://github.com/coala/coala) project to `requests` 2.12.1 we encounter an exception in our test suites which seems to be caused by `requests`.

Build: https://ci.appveyor.com/project/coala/coala-bears/build/1.0.3537/job/1wm7b4u9yhgkxkgn

Relevant part:
```
================================== FAILURES ===================================
_________________ InvalidLinkBearTest.test_redirect_threshold _________________
self = <tests.general.InvalidLinkBearTest.InvalidLinkBearTest testMethod=test_redirect_threshold>
    def test_redirect_threshold(self):
    
        long_url_redirect = """
            https://bitbucket.org/api/301
            https://bitbucket.org/api/302
            """.splitlines()
    
        short_url_redirect = """
            http://httpbin.org/status/301
            """.splitlines()
    
        self.assertResult(valid_file=long_url_redirect,
                          invalid_file=short_url_redirect,
>                         settings={'follow_redirects': 'yeah'})
tests\general\InvalidLinkBearTest.py:157: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests\general\InvalidLinkBearTest.py:75: in assertResult
    out = list(uut.run("valid", valid_file, **settings))
bears\general\InvalidLinkBear.py:80: in run
    file, timeout, link_ignore_regex):
bears\general\InvalidLinkBear.py:53: in find_links_in_file
    code = InvalidLinkBear.get_status_code(link, timeout)
bears\general\InvalidLinkBear.py:37: in get_status_code
    timeout=timeout).status_code
C:\Python34\lib\site-packages\requests\api.py:96: in head
    return request('head', url, **kwargs)
C:\Python34\lib\site-packages\requests\api.py:56: in request
    return session.request(method=method, url=url, **kwargs)
C:\Python34\lib\site-packages\requests\sessions.py:488: in request
    resp = self.send(prep, **send_kwargs)
C:\Python34\lib\site-packages\requests_mock\mocker.py:69: in _fake_send
    return self._real_send(session, request, **kwargs)
C:\Python34\lib\site-packages\requests\sessions.py:641: in send
    r.content
C:\Python34\lib\site-packages\requests\models.py:772: in content
    self._content = bytes().join(self.iter_content(CONTENT_CHUNK_SIZE)) or bytes()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
    def generate():
        # Special case for urllib3.
        if hasattr(self.raw, 'stream'):
            try:
                for chunk in self.raw.stream(chunk_size, decode_content=True):
                    yield chunk
            except ProtocolError as e:
                raise ChunkedEncodingError(e)
            except DecodeError as e:
                raise ContentDecodingError(e)
            except ReadTimeoutError as e:
                raise ConnectionError(e)
        else:
            # Standard file-like object.
            while True:
>               chunk = self.raw.read(chunk_size)
E               AttributeError: 'NoneType' object has no attribute 'read'
C:\Python34\lib\site-packages\requests\models.py:705: AttributeError
```
happens on Windows and Linux.

Thanks in advance :)

### PR DESCRIPTION
**Pull request by nsoranzo**
Fix #3698.

This fixes also a traceback in https://github.com/galaxyproject/bioblend/blob/master/bioblend/galaxy/client.py#L135 where the `text` attribute of an empty `requests.Response()` object is used:
```
Traceback (most recent call last):
  File "/usr/users/ga002/soranzon/software/galaxyproject_bioblend/tests/TestGalaxyInstance.py", line 37, in test_get_retry
    self.gi.libraries.get_libraries()
  File "/usr/users/ga002/soranzon/software/galaxyproject_bioblend/bioblend/galaxy/libraries/__init__.py", line 218, in get_libraries
    libraries = self._get(deleted=deleted)
  File "/usr/users/ga002/soranzon/software/galaxyproject_bioblend/bioblend/galaxy/client.py", line 135, in _get
    raise ConnectionError(msg, body=r.text,
  File "/usr/users/ga002/soranzon/software/galaxyproject_bioblend/.tox/py27/local/lib/python2.7/site-packages/requests/models.py", line 796, in text
    if not self.content:
  File "/usr/users/ga002/soranzon/software/galaxyproject_bioblend/.tox/py27/local/lib/python2.7/site-packages/requests/models.py", line 772, in content
    self._content = bytes().join(self.iter_content(CONTENT_CHUNK_SIZE)) or bytes()
  File "/usr/users/ga002/soranzon/software/galaxyproject_bioblend/.tox/py27/local/lib/python2.7/site-packages/requests/models.py", line 705, in generate
    chunk = self.raw.read(chunk_size)
AttributeError: 'NoneType' object has no attribute 'read'
```

### CODE REVIEW

**Comment by Lukasa** [comment]
Heh, I was going to say that we shouldn't tolerate silly status codes until I saw that we already tolerate `0`. That's really quite odd, but fair enough, I guess this is a logical extension of what we already do.

**Comment by Lukasa** [comment]
However, before we merge this, can I ask you to please add a test for this logic?

**Comment by nsoranzo** [comment]
@Lukasa Do you mean a test that for an empty `requests.Response()` object, the `content` property is None?

**Comment by Lukasa** [comment]
Yep. =)

**Comment by nateprewitt** [comment]
A couple of things here just for future readers. This unfortunately wouldn't fix the problem #3698 had because their mocking still required a valid status code in their empty response.

The other thing to think about is that with this patch, empty responses that used to work with valid bodies but no status code will now return `None`. Likely a rare edge case, but #3698 was thought to be unlikely too.

**Comment by Lukasa** [comment]
I remain really unsure that we should tolerate content with no status codes, but clearly we have done it before so... 😕

**Comment by nateprewitt** [comment]
This patch seems completely logical, I don't see a particular reason to special case 0. Just noting the effects for anyone who hits this later.

**Comment by nateprewitt** [comment]
Sorry, going to redact my previous endorsement for a moment.

This doesn't seem like an unreasonable loosening of checks, but it could end up being a bit possible stumbling block. If you forget to set status code on a Response while mocking, you'll now get unexpected content. These are both pretty rare edge cases but I'm not sure if favoring a new one over the old one will be especially productive. The special cased 0 is an conscious opt-in to `None` rather than a default opt-out of being able to get content from a response.

**Comment by kennethreitz** [comment]
I like the proposed logic.

**Comment by nateprewitt** [comment]
Just another passing thought, while we're here. If we're going to loosen this up, should we also broaden the check to `or self.raw is None` so we're solving #3698? Otherwise we still have an awkward failure mode in the event of a valid status_code but no response body.

**Comment by sigmavirus24** [comment]
Honestly, I really don't think we need to solve for #3698.

There's little in our documentation about creating custom Response objects and while Transport Adapters can do that, we don't document what they need to set. With that in mind, if someone's mocking out requests, they should probably be putting more effort into it than this Pull Request makes them. Why? Because that's not how we are expecting users to use the Response object. There are also *several* libraries for doing this, and if someone's using `mock` they can autospec a Response object.

In short, there's little value in this patch to the critical path of Requests.

**Comment by sigmavirus24** [comment]
> but clearly we have done it before so... 😕

I don't agree that because we've done something before we need to continue down that path. I especially feel we don't need to do this when we recognize that it was a mistake to do this previously.

**Comment by nsoranzo** [comment]
@nateprewitt I'd be happy to change the test to:
```
if self.status_code == 0 or self.raw is None:
```
or anything that makes:
```
r = requests.Response()
r.content
```
return `None`. I think most programmers would expect this and that the traceback is a regression.

**Comment by Lukasa** [comment]
> I think most programmers would expect this and that the traceback is a regression.

Would most programmers expect this? I'm not sure that I would.

I think the reality here is that we have discovered that a concern originally raised in #3607 was correct: our defense against `response.raw` being `None` was there not so much because we needed it but because downstream libraries were patching our objects in such a way that `response.raw` was `None`.

From my perspective then, there are two sensible choices. Number one is to change this patch such that we tolerate the *actual problem*: namely, that `response.raw` is `None`. Number two is to decide that we simply don't support `response.raw` being `None`, and the error is exactly correct.

I think I'm leaning towards number one, if only because we've handled this kind of problem in the past (admittedly with an overbroad `except` block).

**Comment by nsoranzo** [comment]
@Lukasa I'll try to explain why I think that this is the expected behaviour:

- `requests.Response` class is clearly part of the `requests` API and creating such objects is useful when dealing with connection exceptions
- the only `__init__` method of `requests.Response()` has no parameters
- using a property of a legitimately created object shouldn't raise an error, unless there is a very good reason

>  Number one is to change this patch such that we tolerate the actual problem: namely, that response.raw is None.

So, is
```
if self.status_code == 0 or self.raw is None:
```
a good solution for you?

**Comment by sigmavirus24** [comment]
> I think I'm leaning towards number one, if only because we've handled this kind of problem in the past (admittedly with an overbroad except block).

This makes sense to me. We can add the requirement that raw not be None in 3.0

@nsoranzo adding the explicit check for None works for me.

**Comment by nsoranzo** [comment]
@sigmavirus24 I've added the explicit check for None and also a test as requested by @Lukasa.

**Comment by Lukasa** [comment]
The explicit check for `None` works. In an ideal world I'd rather change the logic to just check `raw` directly (presumably the status code `0` behaviour is meant to be a crutch for this), but I'd worry about regressing some other wacky edge case. So let's just accept that this is a mess and move on. =P

**Comment by Lukasa** [comment]
Arg, github had refreshed, but not quite enough. =(
